api/routes/join_suggestion.go

```diff
@@ -88,45 +98,92 @@ func JoinSuggestionHandler(esCtor model.MetadataStorageCtor, metaCtors map[strin
-		for _, dataset := range datasets {
-			if !hasSuggestions(dataset) {
+		datamartDatasets := append(datasetsMap[datamart.ProvenanceISI], datasetsMap[datamart.ProvenanceNYU]...)
+		datasets = filterDatasets(datamartDatasets, hasSuggestions)
+
+		localDatasets := make(map[string]*model.Dataset)
+		for provenance, datasets := range datasetsMap {
+			if provenance == datamart.ProvenanceISI || provenance == datamart.ProvenanceNYU {
 				continue
 			}
+			for _, dataset := range datasets {
+				localDatasets[dataset.ID] = dataset
+			}
+		}
 
-			existing, ok := exists[dataset.ID]
-			if !ok {
-				// we don't have it, add it
-				exists[dataset.ID] = dataset
-			} else {
-				// we already have it, if it is `dataset`, replace it
-				if existing.Provenance == datamart.ProvenanceNYU || existing.Provenance == datamart.ProvenanceISI {
-					exists[dataset.ID] = dataset
+		// If a dataset already exists in the local, use the local dataset augmented with join suggestions from the corresponding datamart dataset
+		// Note: there could be multiple nyu datamart result with same dataset id with diffrent join suggestions/score
+		for i := 0; i < len(datasets); i++ {
+			dataset := datasets[i]
+			if locDataset, ok := localDatasets[dataset.ID]; ok {
+				// make a copy of local dataset
+				localDataset := *locDataset
+				localDataset.JoinScore = datasets[i].JoinScore
+				localDataset.JoinSuggestions = datasets[i].JoinSuggestions
+
+				// Column names are normalized while dataset is being ingest
+				// So we retreive normalized column names from the local dataset
+				// using original names from the datamart dataset
+				for _, suggestion := range localDataset.JoinSuggestions {
+					for j, colName := range suggestion.JoinColumns {
+						colNameTokens := strings.Split(colName, ", ")
+						var localColNameTokens []string
+						for _, token := range colNameTokens {
+							// displayName holds the column name of the original datamart dataset
+							localColNameTokens = append(localColNameTokens, getColNameByDisplayName(localDataset, token))
+						}
+						localColName := strings.Join(localColNameTokens, ", ")
+						suggestion.JoinColumns[j] = localColName
+					}
 				}
+				// Some imported local datasets are missing description. In that case, add a description
+				// (I guess this happens because a downloaded dataset from datamart which is being imported and ingested to the local system,
+				// sometimes comes with datasetDoc.json file with no description in it)
+				if localDataset.Description == "" {
+					localDataset.Description = datasets[i].Description
+				}
+				datasets[i] = &localDataset
 			}
 		}
 
-		var deconflicted []*model.Dataset
-		for _, dataset := range exists {
-			deconflicted = append(deconflicted, dataset)
-		}
+		// sort by join score and name
+		sort.Slice(datasets, func(i, j int) bool {
+			return datasets[i].JoinScore > datasets[j].JoinScore || datasets[i].Name < datasets[j].Name
+		})
```

Comment:
Im not sure this sort works as if join score of one dataset is > then the other but also > for the name then true is returned for both OR clauses. A quick test on https://play.golang.org/p/WG88-O--VD- can illustrate it a bit.

Reply:
I think you are correct. Thanks for catching this. I'll fix!